### PR TITLE
Andrei Tcibin Part3

### PR DESCRIPTION
**Pull request by tcibinan**

### CODE REVIEW

**Comment by tcibinan** [comment]
pull request update:
- added Traversable
- added ReachIterable (with fixed object inconsistency after a several calls of match methods)

File: src/test/java/lambda/part3/exercise/Mapping.java [hunk 1]
```diff
@@ -108,65 +105,108 @@ public void mapping() {
...
+            return result;
         }
 
-        // TODO filter
         // (T -> boolean) -> (T -> [T])
         // filter: [T1, T2] -> (T -> boolean) -> [T2]
         // flatMap": [T1, T2] -> (T -> [T]) -> [T2]
 
+        public LazyFlatMapHelper<T, R> filter(Predicate<T> predicate) {
+            return new LazyFlatMapHelper<>(list, item -> {
```
Comment (senia-psm) [inline src/test/java/lambda/part3/exercise/Mapping.java]: You can call `flatMap` here.

File: src/test/java/lambda/part3/exercise/Mapping.java [hunk 2]
```diff
@@ -224,4 +262,418 @@ public void lazy_mapping() {
...
+
+        assertEquals(
+                Arrays.asList(4, 5),
+                Traversable.from(dataList).filter(i -> i > 3).force()
+        );
+    }
+
+    interface ReachIterable<T> {
+
+        void resetIteration();
```
Comment (senia-psm) [inline src/test/java/lambda/part3/exercise/Mapping.java]: It's an overkill. After all `ReachIterable` should be similar to `Iterable`

File: src/test/java/lambda/part3/exercise/Mapping.java [hunk 3]
```diff
@@ -224,4 +262,418 @@ public void lazy_mapping() {
...
+
+            return new ReachIterable<T>() {
+                @Override
+                public void resetIteration() {
+                    self.resetIteration();
+                }
+
+                @Override
+                public boolean forNext(Consumer<T> c) {
+                    return self.forNext(t -> {
```
Comment (senia-psm) [inline src/test/java/lambda/part3/exercise/Mapping.java]: Incorrect. Non empty `ReachIterable` should call `Consumer::accept` in `forNext`.
This implementation can return `true` without calling `accept`.

File: src/test/java/lambda/part3/exercise/Mapping.java [hunk 4]
```diff
@@ -224,4 +262,418 @@ public void lazy_mapping() {
...
+
+            return new ReachIterable<R>() {
+                @Override
+                public void resetIteration() {
+                    self.resetIteration();
+                }
+
+                @Override
+                public boolean forNext(Consumer<R> c) {
+                    return self.forNext(t ->
```
Comment (senia-psm) [inline src/test/java/lambda/part3/exercise/Mapping.java]: Incorrect. `ReachIterable` can call `Consumer` only once.

File: src/test/java/lambda/part3/exercise/Mapping.java [hunk 5]
```diff
@@ -224,4 +262,418 @@ public void lazy_mapping() {
...
+                            f.apply(t).forEach(c)
+                    );
+                }
+            };
+        }
+
+        default boolean anyMatch(Predicate<T> p) {
+            final boolean[] found = {false};
+            final boolean[] isNotEnd = {true};
```
Comment (senia-psm) [inline src/test/java/lambda/part3/exercise/Mapping.java]: You don't need array for `isNotEnd`.

File: src/test/java/lambda/part3/exercise/Mapping.java [hunk 6]
```diff
@@ -224,4 +262,418 @@ public void lazy_mapping() {
...
+                    }
+                });
+            }
+
+            resetIteration();
+            return found[0];
+        }
+
+        default boolean allMatch(Predicate<T> p) {
+            final boolean[] isAllMatch = {true};
```
Comment (senia-psm) [inline src/test/java/lambda/part3/exercise/Mapping.java]: You can implement `allMatch` using `anyMatch`

File: src/test/java/lambda/part3/exercise/Mapping.java [hunk 7]
```diff
@@ -224,4 +262,418 @@ public void lazy_mapping() {
...
+            resetIteration();
+            return isAllMatch[0];
+        }
+
+        default boolean nonMatch(Predicate<T> p) {
+            if (anyMatch(p)) {
+                return false;
+            }
+
+            final boolean[] isNonMatch = {true};
```
Comment (senia-psm) [inline src/test/java/lambda/part3/exercise/Mapping.java]: You don't need this. Just add `else` branch to `if` above.

**Comment by tcibinan** [comment]
pull request update:

- fixed ReachIterable filter, flatMap and matches methods
- remove consistency tests from ReachIterable (ReachIterable is single time use now such as Iterable)

File: src/test/java/lambda/part3/exercise/Mapping.java [hunk 8]
```diff
@@ -405,83 +398,63 @@ public boolean forNext(Consumer<R> c) {
...
 
                 @Override
                 public boolean forNext(Consumer<R> c) {
-                    return self.forNext(t ->
-                            f.apply(t).forEach(c)
-                    );
+                    boolean result = currentIterable.forNext(c);
+
+                    if (!result) {
+                        boolean isNotEnd = self.forNext(getNewIterable);
```
Comment (senia-psm) [inline src/test/java/lambda/part3/exercise/Mapping.java]: Incorrect. Add test like this: `Arrays.asList("", "", "", "a")` -> `flatMap(getChars).force()`.

File: src/test/java/lambda/part3/exercise/Mapping.java [hunk 9]
```diff
@@ -602,78 +569,4 @@ public void testReachIterableFirstMatch() {
         assertFalse(optional.isPresent());
     }
 
-    @Test
```
Comment (senia-psm) [inline src/test/java/lambda/part3/exercise/Mapping.java]: You need tests.

File: src/test/java/lambda/part3/exercise/Mapping.java [hunk 10]
```diff
@@ -407,12 +407,11 @@ public boolean forNext(Consumer<R> c) {
...
                 public boolean forNext(Consumer<R> c) {
                     boolean result = currentIterable.forNext(c);
+                    boolean isNotEnd = true;
 
-                    if (!result) {
-                        boolean isNotEnd = self.forNext(getNewIterable);
-                        if (isNotEnd) {
-                            result = currentIterable.forNext(c);
-                        }
+                    while (!result && isNotEnd) {
```
Comment (senia-psm) [inline src/test/java/lambda/part3/exercise/Mapping.java]: Incorrect. Add a test with `Arrays.asList("", "ab", "")`

Comment (tcibinan) [reply src/test/java/lambda/part3/exercise/Mapping.java]: on [line 525](https://github.com/java8-course/lambda/pull/120/files/9bdebd0f9eff56aab3c93cd2ed8e0975f1bf2b4c#diff-813ac2502b8b33c07ee7111af3dd7543R525) where is such a test. And I also have tested different combinations of empty strings with flatMap. Can't get a clue what is wrong with this method.

Comment (senia-psm) [reply src/test/java/lambda/part3/exercise/Mapping.java]: Note `ab`, not `a`.

**Comment by NameOfTheLaw** [comment]
test goes green on `Arrays.asList("", "ab", "")`

File: src/test/java/lambda/part3/exercise/Mapping.java (hunk 10 above)
Comment (tcibinan) [reply src/test/java/lambda/part3/exercise/Mapping.java]: test goes green on `Arrays.asList("", "ab", "")`

Comment (senia-psm) [reply src/test/java/lambda/part3/exercise/Mapping.java]: I missed line 409. Yes, it looks correct.
